api/src/main/java/org/apache/iceberg/DistributionMode.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+/**
+ * Enum of supported write distribution mode, it defines the write behavior of batch or streaming job:
+ * <p>
+ * 1. none: don't shuffle rows. It is suitable for scenarios where the rows are located in only few
+ * partitions, otherwise that may produce too many small files because each task is writing rows into different
+ * partitions randomly.
+ * <p>
+ * 2. hash-partition: hash distribute by partition keys, which is suitable for the scenarios where the rows are located
+ * into different partitions evenly.
+ * <p>
+ * 3. range-partition: range distribute by partition key (or sort key if table has an {@link SortOrder}), which is
+ * suitable for the scenarios where rows are located into different partitions with skew distribution.
+ */
+public enum DistributionMode {
+  NONE("none"), HASH("hash-partition"), RANGE("range-partition");
```

Comment:
As I noted in the comment thread, I think that we should use "partition" to describe only table partitions. Otherwise, we are going to create confusion. We can use "hash" and "range" here if there is consensus that "partition" and "sort" are not clear, but I don't think that we should use the term "partition" to refer to distribution within a processing engine.

Reply:
"Partition" and "sort" aren't very clear to me. Both "hash partition" and "range partition" are "partitions". But in the table, they are listed as "partition" and "sort".

If Flink and Spark are going to support different behaviors. Maybe it is better for them to define different values to be more accurately describe the behavior.

Reply:
+1 on not using the term "partition" when talking about distribution.

W.r.t. naming, I did ask myself the question what should be the best names here. I tend to like "hash" and "range" a bit more as it may not be clear that `partition` refers to the table's partition spec.

I guess the real question here is what does this table control? Are we allowing to control whether to use hash or range distribution or do we control whether the distribution is based on the partition spec or sort order?

Reply:
I think it controls whether we use hash distribution or range distribution. I agree that's more clear from a developer's perspective. My concern is that users won't know what hash and range are, but they do understand what a partition is and what sorting is.

Let's go with hash and range for now. I think we can explain it well enough in docs, and we can also add aliases that are more clear if needed.

Reply:
Sounds good to me.

Reply:
what if we call config name as `write.shuffle-mode`? would it make it more clear to user regarding `hash` vs `range`?

Reply:
@stevenzwu ,   I think `write.shuffle-mode` is enough to express the write behavior of flink, but not enough to express the write behavior of spark, because spark will distribute those records with local sort or global sort. 

+1 on keeping `write.distribution-mode` and using the `hash` & `range` values now ( Though the `range` does not fully express the sort meaning from spark, but I can not think of a better word to express the exact meaning for both flink and spark ).